### PR TITLE
mesa_23: init at 23.0.1

### PR DESCRIPTION
**Pull request by Atemu**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Atemu** [comment]
cc @hw-lunemann

**Comment by vcunat** [comment]
Well, _if_ we're going to do this, we should describe some way how to use it, right?  (I think it will have more value later, on stable NixOS)

I know there are options `hardware.opengl.package` (and `.package32`) but they're marked as internal and I'm not sure if switching this way really works without issues.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -22225,12 +22225,15 @@ with pkgs;
   # Default libGLU
   libGLU = mesa_glu;
 
-  mesa = callPackage ../development/libraries/mesa {
-    llvmPackages = llvmPackages_15;
-    stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
+  mesa_22 = darwin.apple_sdk_11_0.callPackage ../development/libraries/mesa/22.nix {
```
Comment (andrevmatos) [inline pkgs/top-level/all-packages.nix:22228]: why `apple_sdk_11_0` including for linux envs?

**Comment by Atemu** [comment]
@vcunat because `package` is also used for switching out the OGL implementation to something else entirely; i.e. AMD's proprietary driver.

Mesa has its own specialities like the `drivers` output.

**Comment by vcunat** [comment]
I don't see the difference (yet).  Both is switching to a different libGL (and friends).  If anything, it should fail by default if both internal code sets it (because some other option) and user sets it.

**Comment by K900** [comment]
I think what they're saying is that Mesa needs to specifically choose the `drivers` output and other drivers don't, but I think this shouldn't really be a problem - we can just tell the users to use `mesa.drivers`.

**Comment by K900** [comment]
Not sure I like that. Sounds really easy to break. I really wish we had a generic "provides" mechanism in nixpkgs sometimes...

**Comment by vcunat** [comment]
It could be improved fairly easily, but there just a couple particular packages that make sense to pass, and even generally I wouldn't expect `.drivers` to exist commonly.

**Comment by Atemu** [comment]
> even generally I wouldn't expect `.drivers` to exist commonly.

That's sort of why I wanted a mesa-specific option.

I get your points but I still think it's better to provide an option that caters towards mesa specifically as an interface to the end-user.

**Comment by L911P** [comment]
you'll want to apply the fix from https://github.com/NixOS/nixpkgs/pull/219802 because otherwise with mesa 23 intel devices won't have 32bit vulkan drivers.

**Comment by cole-h** [comment]
> I know there are options `hardware.opengl.package` (and `.package32`) but they're marked as internal and I'm not sure if switching this way really works without issues.

Just as a data point, I've been able to use this "internal" option just fine: https://github.com/cole-h/nixos-config/blob/3d8ba5971794df6422353c5236e7c06bb36428de/hosts/scadrial/modules/rtx4090-compat.nix#L29-L38. It just doesn't appear in `man configuration.nix`.

**Comment by K900** [comment]
23.0.1 is finally out! https://gitlab.freedesktop.org/mesa/mesa/-/tags/mesa-23.0.1

File: nixos/modules/hardware/opengl.nix [hunk 2]
```diff
@@ -69,21 +69,48 @@ in
...
-          set.
+          Same as {option}`package` but for the 32-bit OpenGL implementation on
+          64-bit systems. Used when {option}`driSupport32Bit` is set.
         '';
       };
 
+      mesaPackage = mkOption {
+        type = types.package;
+        default = pkgs.mesa;
+        example = "pkgs.mesa_23";
```
Comment (Scrumplex) [inline nixos/modules/hardware/opengl.nix:94]:
```suggestion
        default = pkgs.mesa;
        defaultText = literalExpression "pkgs.mesa";
        example = literalExpression "pkgs.mesa_23";
```

File: nixos/modules/hardware/opengl.nix [hunk 3]
```diff
@@ -69,21 +69,48 @@ in
...
+
+          You should only need to adjust this if you require a newer mesa
+          version for your hardware or because you need to patch a bug.
+        '';
+        apply = mesa: mesa.drivers or throw "mesa package must have a `drivers` output.";
+      };
+      mesaPackage32 = mkOption {
+        type = types.package;
+        default = pkgs.mesa;
+        example = "pkgs.pkgsi686Linux.mesa_23";
```
Comment (Scrumplex) [inline nixos/modules/hardware/opengl.nix:106]:
```suggestion
        default = pkgs.pkgsi686Linux.mesa;
        defaultText = literalExpression "pkgs.pkgsi686Linux.mesa";
        example = literalExpression "pkgs.pkgsi686Linux.mesa_23";
```

File: nixos/modules/hardware/opengl.nix [hunk 4]
```diff
@@ -69,21 +69,50 @@ in
       package = mkOption {
         type = types.package;
         internal = true;
+        default = cfg.mesaPackage;
         description = lib.mdDoc ''
           The package that provides the OpenGL implementation.
+
+          The default is mesa's drivers which should cover all OpenGL-capable
```
Comment (K900) [inline nixos/modules/hardware/opengl.nix:76]: Upstream spells Mesa with a capital M. We should too, probably.

File: nixos/modules/hardware/opengl.nix [hunk 5]
```diff
@@ -69,21 +69,50 @@ in
...
         '';
       };
 
+      mesaPackage = mkOption {
+        type = types.package;
+        default = pkgs.mesa_23;
+        defaultText = literalExpression "pkgs.mesa_23";
+        example = literalExpression "pkgs.mesa_22";
+        description = lib.mdDoc ''
+          The mesa driver package used for rendering support on the system.
```
Comment (K900) [inline nixos/modules/hardware/opengl.nix:97]: Nit: rendering -> rendering acceleration?

File: nixos/modules/hardware/opengl.nix [hunk 6]
```diff
@@ -69,21 +69,50 @@ in
...
+
+          You should only need to adjust this if you require a newer mesa
```
Comment (K900) [inline nixos/modules/hardware/opengl.nix:99]: Nit: not really the case? You can totally want to update for newer features and such.

**Review by K900 — CHANGES_REQUESTED** [review]
Some documentation nits.

**Comment by K900** [comment]
I have no idea why Github is not letting me comment on that line, but `hardware.opengl.package` is also set further down in `opengl.nix` and that should probably be removed if we're adding a real default.

File: nixos/modules/hardware/opengl.nix (hunk 5 above)
Comment (Atemu) [reply nixos/modules/hardware/opengl.nix:97]: If we're splitting hairs, Mesa has non-accelerated rendering support too ;)

Comment (K900) [reply nixos/modules/hardware/opengl.nix:97]: Honestly, I'd probably go for something like "to provide graphics drivers", even.

File: nixos/modules/hardware/opengl.nix (hunk 6 above)
Comment (Atemu) [reply nixos/modules/hardware/opengl.nix:99]: Reasoning is that if you require newer features, you require a new mesa in my view.

**Comment by Atemu** [comment]
Good catch!

File: nixos/modules/hardware/opengl.nix [hunk 7]
```diff
@@ -69,19 +69,48 @@ in
...
         '';
+        apply = mesa: mesa.drivers or throw "`mesa` package must have a `drivers` output.";
+      };
+      mesaPackage32 = mkOption {
+        type = types.package;
+        default = pkgs.pkgsi686Linux.mesa_23;
+        defaultText = literalExpression "pkgs.pkgsi686Linux.mesa_23";
+        example = literalExpression "pkgs.pkgsi686Linux.mesa_22";
+        description = lib.mdDoc ''
+          Same as {option}`mesaPackage` but for the 32-bit mesa on 64-bit
```
Comment (K900) [inline nixos/modules/hardware/opengl.nix:110]: One last lower case "mesa" here ^^"

**Review by K900 — APPROVED** [review]
LGTM!

**Comment by vcunat** [comment]
At some point we might add a bullet into 23.05 release notes, mentioning that you can switch the driver versions.

The editing tasks around mesa releases got way more complicated by this, needing file creation and (possibly) editing a few of them, but I suppose it's OK.  I know I'm a bit late with looking more properly at this; overall I do like this PR.  Thanks.

**Comment by Atemu** [comment]
Looking at [22.11's release notes](https://nixos.org/manual/nixos/stable/release-notes.html#sec-release-22.11), something like this could indeed be mentioned. Perhaps even as a candidate for a release highlight.

**Comment by K900** [comment]
I was thinking we should do that in the staging change that actually updates the default.

**Comment by vcunat** [comment]
Could be, but I don't see any need to couple these two.

**Comment by maxeaubrey** [comment]
Not sure if it's just me/my procrastination on switching to flakes, but I cannot rebuild my system with bf1785d0a91 in.

```
error: attempt to call something which is not a function but a set

       at /home/systems/nixpkgs/nixos/modules/hardware/opengl.nix:102:23:

          101|         '';
          102|         apply = mesa: mesa.drivers or throw "`mesa` package must have a `drivers` output.";
             |                       ^
          103|       };
```

**Comment by K900** [comment]
Can you run that with `--show-trace`? Is `hardware.opengl.package` set in your config?

**Comment by maxeaubrey** [comment]
@K900 I do not set `hardware.opengl.package` explicitly anywhere in my config. Sorry about the lack of a full trace, here you go: https://gist.github.com/maxeaubrey/aff612ba133fbfbf28f86c4820edbfd5

**Comment by Atemu** [comment]
Found the error, fix incoming.

**Comment by vcunat** [comment]
For reference: https://github.com/NixOS/nixpkgs/commit/1d19eeefeea4cb51d96592712acc4443c42efd0b

**Pull request merged**